msal_extensions/_windows.py

```diff
@@ -0,0 +1,173 @@
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_protect_data(
+                ctypes.byref(message_blob),
+                u"python_data",
+                entropy,
+                None,
+                None,
+                _CRYPTPROTECT_UI_FORBIDDEN,
+                ctypes.byref(result)):
+            return result.raw()
+        return b''
+
+    def unprotect(self, cipher_text):
+        # type: (bytes) -> str
+        ct_buffer = ctypes.c_buffer(cipher_text, len(cipher_text))
+        ct_blob = DATA_BLOB(len(cipher_text), ct_buffer)
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_unprotect_data(
+            ctypes.byref(ct_blob),
+            None,
+            entropy,
+            None,
+            None,
+            _CRYPTPROTECT_UI_FORBIDDEN,
+            ctypes.byref(result)
+        ):
+            return result.raw().decode('utf-8')
+        return u''
+
+
+class _WindowsTokenCache(msal.SerializableTokenCache):
+    DEFAULT_CACHE_LOCATION = os.path.join(os.getenv('LOCALAPPDATA'), '.IdentityService', 'msal.cache')
+    DEFAULT_ENTROPY = ''
+
+    def __init__(self, **kwargs):
+        super(_WindowsTokenCache, self).__init__()
+
+        self._cache_location = _WindowsTokenCache.DEFAULT_CACHE_LOCATION  # type: str
+        if 'cache_location' in kwargs:
+            self._cache_location = kwargs['cache_location'] or _WindowsTokenCache.DEFAULT_CACHE_LOCATION
+        self._lock_location = self._cache_location + '.lockfile'
+
+        entropy = _WindowsTokenCache.DEFAULT_ENTROPY
+        if 'entropy' in kwargs:
+            entropy = kwargs['entropy']
+        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
+        self._last_sync = 0  # _last_sync is a Unixtime
+
+    def _has_state_changed(self):
+        # type: () -> Bool
+        try:
+            return self.has_state_changed or self._last_sync < os.path.getmtime(self._cache_location)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise e
+            return False
+
+    def add(self, event, **kwargs):
+        if self._has_state_changed:
+            try:
+                self._read()
+            except OSError as e:
+                if e.errno != errno.ENOENT:
+                    raise e
+        super(_WindowsTokenCache, self).add(event, **kwargs)
+        self._write()
+
+    def update_rt(self, rt_item, new_rt):
+        if self.has_state_changed:
+            try:
+                self._read()
+            except OSError as e:
+                if e.errno != errno.ENOENT:
+                    raise e
+        super(_WindowsTokenCache, self).update_rt(rt_item, new_rt)
+        self._write()
+
+    def remove_rt(self, rt_item):
```

Comment:
Same as above `update_rt()`.

Reply:
Same as here: https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/14#discussion_r286144472